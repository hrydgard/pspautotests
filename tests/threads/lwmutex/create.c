#include "shared.h"
#include <limits.h>

#define CREATE_TEST(title, workarea, name, attr, count, options) { \
	int result = sceKernelCreateLwMutex(workarea, name, attr, count, options); \
	if (result == 0) { \
		checkpoint(NULL); \
		schedf("%s: OK ", title); \
		schedfLwMutex(workarea); \
		sceKernelDeleteLwMutex(workarea); \
	} else { \
		checkpoint("%s: Failed (%X)", title, result); \
	} \
}

int main(int argc, char **argv) {
	SceLwMutexWorkarea workarea, workarea1, workarea2;
	int result1, result2;
	int i;
	u32 options[32];
	memset(options, 0, sizeof(options));

	// Crash.
	//CREATE_TEST("NULL workarea", NULL, "test", 0, 0, NULL);

	checkpointNext("Names:");
	CREATE_TEST("  NULL name", &workarea, NULL, 0, 0, NULL);
	CREATE_TEST("  Blank name", &workarea, "", 0, 0, NULL);
	CREATE_TEST("  Long name", &workarea, "1234567890123456789012345678901234567890123456789012345678901234", 0, 0, NULL);

	result1 = sceKernelCreateLwMutex(&workarea1, "create", 0, 0, NULL);
	result2 = sceKernelCreateLwMutex(&workarea2, "create", 0, 0, NULL);
	if (result1 == 0 && result2 == 0) {
		checkpoint("  Two with same name: OK\n");
	} else {
		checkpoint("  Two with same name: Failed (%X, %X)\n", result1, result2);
	}
	sceKernelDeleteLwMutex(&workarea1);
	sceKernelDeleteLwMutex(&workarea2);

	checkpointNext("Attributes:");
	const static int attrs[] = {1, 0x100, 0x200, 0x300, 0x3FF, 0x400, 0x800, 0xB00, 0x1000, 0x2000, 0x4000, 0x8000, 0x10000};
	for (i = 0; i < ARRAY_SIZE(attrs); ++i) {
		char temp[32];
		sprintf(temp, "  %x", attrs[i]);
		CREATE_TEST(temp,  &workarea, "create", attrs[i], 0, NULL);
	}

	checkpointNext("Counts:");
	CREATE_TEST("  Positive count", &workarea, "create", 0, 1, NULL);
	CREATE_TEST("  Negative count", &workarea, "create", 0, -1, NULL);
	CREATE_TEST("  Count = -5", &workarea, "create", 0, -5, NULL);
	CREATE_TEST("  Count = 2", &workarea, "create", 0, 2, NULL);
	CREATE_TEST("  Negative count (recursive)", &workarea, "create", 0x200, -1, NULL);
	CREATE_TEST("  Count = -5 (recursive)", &workarea, "create", 0x200, -5, NULL);
	CREATE_TEST("  Count = 2 (recursive)", &workarea, "create", 0x200, 2, NULL);
	CREATE_TEST("  Medium count (recursive)", &workarea, "create", 0x200, 255, NULL);
	CREATE_TEST("  Large count (recursive)", &workarea, "create", 0x200, 65537, NULL);

	checkpointNext("Option sizes:");
	const static int optionSizes[] = {-1, 0, 1, 4, 8, 16, 32, 64, 256, 0x7FFFFFFF};
	for (i = 0; i < ARRAY_SIZE(optionSizes); ++i) {
		char temp[32];
		sprintf(temp, "  %d", optionSizes[i]);
		options[0] = optionSizes[i];
		CREATE_TEST(temp, &workarea, "create", 0, 1, options);
	}

	checkpointNext("Scheduling:");
	sceKernelCreateLwMutex(&workarea, "create1", 0, 0, NULL);
	CREATE_TEST("Two with same workarea", &workarea, "create2", 0, 0, NULL);
	// Leaks.

	BASIC_SCHED_TEST("NULL name",
		result = sceKernelCreateLwMutex(&workarea, NULL, 0, 0, NULL);
	);
	BASIC_SCHED_TEST("Negative count",
		result = sceKernelCreateLwMutex(&workarea, "create2", 0, -1, NULL);
	);
	BASIC_SCHED_TEST("Create locked",
		result = sceKernelCreateLwMutex(&workarea, "create2", 0, 1, NULL);
	);
	sceKernelDeleteLwMutex(&workarea);
	BASIC_SCHED_TEST("Create not locked",
		result = sceKernelCreateLwMutex(&workarea, "create2", 0, 0, NULL);
	);
	sceKernelDeleteLwMutex(&workarea);

	LOCKED_SCHED_TEST("Initial not locked", 0, 0,
		sceKernelDelayThread(1000);
		result = 0;
	);

	LOCKED_SCHED_TEST("Initial locked", 1, 0,
		sceKernelDelayThread(1000);
		result = 0;
	);

	SceLwMutexWorkarea workareas[1024];
	int result = 0;
	for (i = 0; i < 1024; i++)
	{
		result = sceKernelCreateLwMutex(&workareas[i], "create", 0, 0, NULL);
		if (result != 0)
			break;
	}

	if (result != 0)
		schedf("Create 1024: Failed at %d (%08X)\n", i, result);
	else
		schedf("Create 1024: OK\n");

	while (--i >= 0)
		sceKernelDeleteLwMutex(&workareas[i]);

	return 0;
}