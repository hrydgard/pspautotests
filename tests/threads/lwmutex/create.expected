[x] Names:
[x]   NULL name: Failed (80020001)
[x]   Blank name: OK LwMutex: OK (size=64,name=,attr=00000000,uid=1,workarea=1,init=0,current=0,lockThread=0)
LwMutexWorkarea (count=0, thread=00000000, attr=000, waiting=0, uid=1, 00000000, 00000000, 00000000)
[x]   Long name: OK LwMutex: OK (size=64,name=1234567890123456789012345678901,attr=00000000,uid=1,workarea=1,init=0,current=0,lockThread=0)
LwMutexWorkarea (count=0, thread=00000000, attr=000, waiting=0, uid=1, 00000000, 00000000, 00000000)
[x]   Two with same name: OK


[x] Attributes:
[x]   1: OK LwMutex: OK (size=64,name=create,attr=00000001,uid=1,workarea=1,init=0,current=0,lockThread=0)
LwMutexWorkarea (count=0, thread=00000000, attr=001, waiting=0, uid=1, 00000000, 00000000, 00000000)
[x]   100: OK LwMutex: OK (size=64,name=create,attr=00000100,uid=1,workarea=1,init=0,current=0,lockThread=0)
LwMutexWorkarea (count=0, thread=00000000, attr=100, waiting=0, uid=1, 00000000, 00000000, 00000000)
[x]   200: OK LwMutex: OK (size=64,name=create,attr=00000200,uid=1,workarea=1,init=0,current=0,lockThread=0)
LwMutexWorkarea (count=0, thread=00000000, attr=200, waiting=0, uid=1, 00000000, 00000000, 00000000)
[x]   300: OK LwMutex: OK (size=64,name=create,attr=00000300,uid=1,workarea=1,init=0,current=0,lockThread=0)
LwMutexWorkarea (count=0, thread=00000000, attr=300, waiting=0, uid=1, 00000000, 00000000, 00000000)
[x]   3ff: OK LwMutex: OK (size=64,name=create,attr=000003ff,uid=1,workarea=1,init=0,current=0,lockThread=0)
LwMutexWorkarea (count=0, thread=00000000, attr=3FF, waiting=0, uid=1, 00000000, 00000000, 00000000)
[x]   400: Failed (80020191)
[x]   800: Failed (80020191)
[x]   b00: Failed (80020191)
[x]   1000: Failed (80020191)
[x]   2000: Failed (80020191)
[x]   4000: Failed (80020191)
[x]   8000: Failed (80020191)
[x]   10000: Failed (80020191)

[x] Counts:
[x]   Positive count: OK LwMutex: OK (size=64,name=create,attr=00000000,uid=1,workarea=1,init=1,current=1,lockThread=1)
LwMutexWorkarea (count=1, thread=00000001, attr=000, waiting=0, uid=1, 00000000, 00000000, 00000000)
[x]   Negative count: Failed (800201BD)
[x]   Count = -5: Failed (800201BD)
[x]   Count = 2: Failed (800201BD)
[x]   Negative count (recursive): Failed (800201BD)
[x]   Count = -5 (recursive): Failed (800201BD)
[x]   Count = 2 (recursive): OK LwMutex: OK (size=64,name=create,attr=00000200,uid=1,workarea=1,init=2,current=2,lockThread=1)
LwMutexWorkarea (count=2, thread=00000001, attr=200, waiting=0, uid=1, 00000000, 00000000, 00000000)
[x]   Medium count (recursive): OK LwMutex: OK (size=64,name=create,attr=00000200,uid=1,workarea=1,init=255,current=255,lockThread=1)
LwMutexWorkarea (count=255, thread=00000001, attr=200, waiting=0, uid=1, 00000000, 00000000, 00000000)
[x]   Large count (recursive): OK LwMutex: OK (size=64,name=create,attr=00000200,uid=1,workarea=1,init=65537,current=65537,lockThread=1)
LwMutexWorkarea (count=65537, thread=00000001, attr=200, waiting=0, uid=1, 00000000, 00000000, 00000000)

[x] Option sizes:
[x]   -1: OK LwMutex: OK (size=64,name=create,attr=00000000,uid=1,workarea=1,init=1,current=1,lockThread=1)
LwMutexWorkarea (count=1, thread=00000001, attr=000, waiting=0, uid=1, 00000000, 00000000, 00000000)
[x]   0: OK LwMutex: OK (size=64,name=create,attr=00000000,uid=1,workarea=1,init=1,current=1,lockThread=1)
LwMutexWorkarea (count=1, thread=00000001, attr=000, waiting=0, uid=1, 00000000, 00000000, 00000000)
[x]   1: OK LwMutex: OK (size=64,name=create,attr=00000000,uid=1,workarea=1,init=1,current=1,lockThread=1)
LwMutexWorkarea (count=1, thread=00000001, attr=000, waiting=0, uid=1, 00000000, 00000000, 00000000)
[x]   4: OK LwMutex: OK (size=64,name=create,attr=00000000,uid=1,workarea=1,init=1,current=1,lockThread=1)
LwMutexWorkarea (count=1, thread=00000001, attr=000, waiting=0, uid=1, 00000000, 00000000, 00000000)
[x]   8: OK LwMutex: OK (size=64,name=create,attr=00000000,uid=1,workarea=1,init=1,current=1,lockThread=1)
LwMutexWorkarea (count=1, thread=00000001, attr=000, waiting=0, uid=1, 00000000, 00000000, 00000000)
[x]   16: OK LwMutex: OK (size=64,name=create,attr=00000000,uid=1,workarea=1,init=1,current=1,lockThread=1)
LwMutexWorkarea (count=1, thread=00000001, attr=000, waiting=0, uid=1, 00000000, 00000000, 00000000)
[x]   32: OK LwMutex: OK (size=64,name=create,attr=00000000,uid=1,workarea=1,init=1,current=1,lockThread=1)
LwMutexWorkarea (count=1, thread=00000001, attr=000, waiting=0, uid=1, 00000000, 00000000, 00000000)
[x]   64: OK LwMutex: OK (size=64,name=create,attr=00000000,uid=1,workarea=1,init=1,current=1,lockThread=1)
LwMutexWorkarea (count=1, thread=00000001, attr=000, waiting=0, uid=1, 00000000, 00000000, 00000000)
[x]   256: OK LwMutex: OK (size=64,name=create,attr=00000000,uid=1,workarea=1,init=1,current=1,lockThread=1)
LwMutexWorkarea (count=1, thread=00000001, attr=000, waiting=0, uid=1, 00000000, 00000000, 00000000)
[x]   2147483647: OK LwMutex: OK (size=64,name=create,attr=00000000,uid=1,workarea=1,init=1,current=1,lockThread=1)
LwMutexWorkarea (count=1, thread=00000001, attr=000, waiting=0, uid=1, 00000000, 00000000, 00000000)

[x] Scheduling:
[x] Two with same workarea: OK LwMutex: OK (size=64,name=create2,attr=00000000,uid=1,workarea=1,init=0,current=0,lockThread=0)
LwMutexWorkarea (count=0, thread=00000000, attr=000, waiting=0, uid=1, 00000000, 00000000, 00000000)
NULL name: A1B1C2E1D1F2 (thread=800201B5, main=80020001)
Negative count: A1B1C2E1D1F2 (thread=800201B5, main=800201BD)
Create locked: A1B1C2E1D1F2 (thread=800201B5, main=00000000)
Create not locked: A1B1C2E1D1F2 (thread=800201B5, main=00000000)
Initial not locked: A1B1D2C2E1F1 (thread=00000000, main=00000000)
Initial locked: A1B1C2E2D1F2 (thread=800201B5, main=00000000)
Create 1024: OK
